contracts/vaults/balancer/external/ComposableAuraHelper.sol

```diff
@@ -0,0 +1,206 @@
+                }
+            }
+        }
+    }
+
+    function settleVaultEmergency(
+        BalancerComposableAuraStrategyContext memory context, 
+        uint256 maturity, 
+        bytes calldata data
+    ) external {
+        ComposableRedeemParams memory params = _decodeParamsAndValidate(
+            context.baseStrategy.vaultSettings.emergencySettlementSlippageLimitPercent,
+            data
+        );
+        bool isSingleSidedExit = params.redemptionTrades.length == 0;
+        
+        uint256 bptToSettle = context.baseStrategy._getEmergencySettlementParams({
+            maturity: maturity, 
+            totalPoolSupply: context.oracleContext.virtualSupply
+        });
+
+        /// @notice minAmounts are not required to be passed in by the caller for this strategy vault
+        uint256[] memory minAmounts = context.poolContext._getMinExitAmounts({
+            oracleContext: context.oracleContext,
+            strategyContext: context.baseStrategy,
+            poolClaim: bptToSettle
+        });
+
+        context.poolContext._unstakeAndExitPool(
+            context.stakingContext, bptToSettle, minAmounts, isSingleSidedExit
+        );
+
+        context.baseStrategy.vaultState.totalPoolClaim -= bptToSettle;
+        context.baseStrategy.vaultState.setStrategyVaultState(); 
+
+        emit VaultEvents.EmergencyVaultSettlement(maturity, bptToSettle, 0);
+    }
+
+    function reinvestReward(
+        BalancerComposableAuraStrategyContext calldata context,
+        ReinvestRewardParams calldata params
+    ) external returns (
+        address rewardToken,
+        uint256 amountSold, 
+        uint256 poolClaimAmount
+    ) {
+        StrategyContext memory strategyContext = context.baseStrategy;
+        BalancerComposablePoolContext calldata poolContext = context.poolContext; 
+        ComposableOracleContext calldata oracleContext = context.oracleContext;
+
+        uint256[] memory amounts;
+        (
+            rewardToken, 
+            amountSold,
+            amounts
+        ) = poolContext._executeRewardTrades({
+            strategyContext: strategyContext,
+            rewardTokens: context.stakingContext.rewardTokens,
+            data: params.tradeData
+        });
+
+        // Make sure we are joining with the right proportion to minimize slippage
+        poolContext._getTimeWeightedPrimaryBalance(
+            oracleContext, strategyContext, 0, true // validateOnly = true
+        );
+
+        poolClaimAmount = poolContext._joinPoolAndStake({
+            strategyContext: strategyContext,
+            oracleContext: oracleContext,
+            stakingContext: context.stakingContext,
+            amounts: amounts,
+            /// @notice minBPT is not required to be set by the caller because primaryAmount
+            /// and secondaryAmount are already validated
+            minBPT: params.minPoolClaim      
```

Comment:
I'm confused - does this function take a minBPT parameter or not? Where have primaryAmount and secondaryAmount been validated?

Reply:
primary and secondary are validated in  poolContext._getTimeWeightedPrimaryBalance (line 147). When validateOnly = true, the function only does validation and skips the valuation math.

Reply:
minBPT is an optional parameter